src/query/storages/fuse/src/operations/read/native_rows_fetcher.rs

```diff
@@ -80,16 +85,74 @@ impl<const BLOCKING_IO: bool> RowsFetcher for NativeRowsFetcher<BLOCKING_IO> {
                 .or_insert(HashSet::from([page_idx]));
             row_set.push((prefix, page_idx, idx_within_page));
         }
+
+        // Read blocks in `prefix` order.
         let part_set = part_set
             .into_iter()
+            .sorted_by_key(|(k, _)| *k)
             .map(|(k, v)| {
                 let mut v = v.into_iter().collect::<Vec<_>>();
                 v.sort();
                 (k, v)
             })
-            .collect();
+            .collect::<Vec<_>>();
+        let mut idx_map = HashMap::with_capacity(part_set.len());
+        for (p, pages) in part_set.iter() {
+            for page in pages {
+                idx_map.insert((*p, *page), idx_map.len());
+            }
+        }
+        // part_set.len() = parts_per_thread * max_threads + remain
+        // task distribution:
+        //   Part number of each task   |       Task number
+        // ------------------------------------------------------
+        //    parts_per_thread + 1      |         remain
+        //      parts_per_thread        |   max_threads - remain
+        let num_parts = part_set.len();
+        let parts_per_thread = num_parts / self.max_threads;
+        let remain = num_parts % self.max_threads;
+        let mut tasks = Vec::with_capacity(self.max_threads);
+        // Fetch blocks in parallel.
+        for i in 0..remain {
```

Comment:
duplicated codes with L128-L140.

You can distribute the parts into threads and make the last thread accept the remains.

Reply:
c++ examples:
```
for (size_t thread_idx = 0; thread_idx < num_threads; ++thread_idx)
      {
          size_t begin = num_keys * thread_idx / num_threads;
          size_t end = num_keys * (thread_idx + 1) / num_threads;

           ...
      }
```

Reply:
The num_keys will change for different parts.